internal/edgeapi/backend/factory/factory.go

```diff
@@ -0,0 +1,44 @@
+package factory
+
+import (
+	"go.uber.org/zap"
+	"k8s.io/client-go/tools/record"
+	kubeclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/project-flotta/flotta-operator/internal/common/repository/edgedevice"
+	"github.com/project-flotta/flotta-operator/internal/common/repository/edgedeviceset"
+	"github.com/project-flotta/flotta-operator/internal/common/repository/edgedevicesignedrequest"
+	"github.com/project-flotta/flotta-operator/internal/common/repository/edgeworkload"
+	"github.com/project-flotta/flotta-operator/internal/common/storage"
+	"github.com/project-flotta/flotta-operator/internal/edgeapi/backend"
+	"github.com/project-flotta/flotta-operator/internal/edgeapi/backend/k8s"
+	"github.com/project-flotta/flotta-operator/internal/edgeapi/configmaps"
+	"github.com/project-flotta/flotta-operator/internal/edgeapi/devicemetrics"
+	"github.com/project-flotta/flotta-operator/internal/edgeapi/images"
+	"github.com/project-flotta/flotta-operator/internal/edgeapi/k8sclient"
+)
+
+func Create(initialDeviceNamespace string, client kubeclient.Client, logger *zap.SugaredLogger, eventRecorder record.EventRecorder) backend.EdgeDeviceBackend {
```

Comment:
It is required for `Create` to return that interface, because it's intended to create different types (two) of backends; now it's only one.

Reply:
If you do not return an interface, what would you return? The interface is the generic type of what the factory is creating, currently it's hardcoded to k8s but that will change depending on input parameters (it could be env var or option set during launch)

Reply:
If it is an option or env var what it is stoping you to be explicit when you are creating the object. You want to hide a `switch` or a `if`. ok but personally I don't see the added value of the factory pattern. I makes the code harder to read.

Reply:
fyi https://github.com/golang/go/wiki/CodeReviewComments#interfaces

Reply:
Maybe `api.go` may be moved into the `factory` package if that's GO good practices but to be as much generic as possible, to ensure we can easily add new backends, we do need the factory; it will be the only part of the code to update to support new backends.
The signature of the `Create` will evolve when it will be made generic; currently the method is expecting a `kubeclient`. So when changing the signature, we will have to be generic enough but not too complex

Reply:
There is another aspect to that - there is no reason to have the concrete type `backend` public, which would be required if the factory method had to return a concrete type.
All-in-all the purpose of that package and method is to hide complexity of creating different backends with their dependencies - the method can be as well put in `cmd/httpapi` and called `CreateK8sBackend` to avoid future-proofing. That, IMO, is less clean than the current approach.

Reply:
`CreateK8sBackend` is much cleaner than `Create` returning an interface. At least I know without digging into the code what I got there.
I'm closing this conversation. There is no reason to keep debating about patterns here.